add MiVideo media creation data (#92)

lsloan opened the pull request:
Resolves #92.

## Prerequisite
#143 should merged before this PR.  After it's merged, I would update this PR to use the new `DataSourceTimestamp` class.  (Alternatively, I could make the changes from #143 part of this PR.)

## Review style
I **_humbly_** request that comments on this PR be nested under a single review.  It helps me if the comments are organized this way, especially if the comments request changes in the code.

## Testing
The MiVideo media creation data comes from the Kaltura API.  The application requires the `kaltura_partner_id` and `kaltura_user_secret` properties in the `MIVIDEO` object of `env.hjson`.  (The correct values are available in the `env_lsloan_2020-05-08_hjson.txt` file I uploaded to the `TL Security files` > `Reporting Project (course-inventory)` folder in Box.)

## Changes
* [x] Remove specifics about `umich-its-tl-reports-prod.json` - https://github.com/tl-its-umich-edu/course-inventory/pull/116#discussion_r423092077
* [x] Remove `kaltura_user_secret` placeholder string from `config/env_blank.hjson` and set validation minimum string length of 1 - https://github.com/tl-its-umich-edu/course-inventory/pull/116#discussion_r423104887
* [x] Remove extra blank line in `config/env_blank.hjson` EOF - https://github.com/tl-its-umich-edu/course-inventory/pull/116#discussion_r423105130
* [x] Move `categoriesFullNameIn` value to config - https://github.com/tl-its-umich-edu/course-inventory/pull/116/files/19f1daea47358b58426ea1c98a9a3b0958cead02#diff-a262ba8dd2b6d9b08986f638bb33f9ac
* [x] Replace `while True` with other logic - https://github.com/tl-its-umich-edu/course-inventory/pull/116#discussion_r423130098
* [x] Move `CANVAS_ZOOM_MEETINGS` - https://github.com/tl-its-umich-edu/course-inventory/pull/116#discussion_r423132754

lsloan commented:
Sorry.  Jumped the gun on marking this ready.  I'll deal with the conflicts first, then request reviews again.

ssciolla commented on mivideo/mivideo_extract.py:
```diff
@@ -2,29 +2,37 @@
 
 import pandas as pd
+from KalturaClient import KalturaClient, KalturaConfiguration
+from KalturaClient.Plugins.Core import (
```
Interesting, I haven't seen this import syntax before.

lsloan replied:
I don't like backslash continuation for various reasons.  I've learned that Python is happy with parenthesized expressions that contain line breaks.  I hadn't used this syntax with `import` before, but I thought I'd give it a try.  It worked.

lsloan commented on config/env_schema.hjson:
```diff
@@ -70,9 +70,18 @@
             "type": "object",
             "properties": {
-                "service_account_json_filename": {"type": "string"}
+                "default_last_timestamp": {"type": "string"}, # should be "date-time", but fails!
+                "udp_service_account_json_filename": {"type": "string"},
+                "kaltura_partner_id": {"type": "integer"},
+                "kaltura_user_id": {"type": "string"}, # should be "email", but fails!
```
@ssciolla, you're more familiar with the HJSON validation than I.  Maybe you know something about this…

* I specified `"date-time"` and `"email"` for the field types where appropriate, but the validation still fails, even when I know the value types in the input are correct.
* Validation order seems to be random.  Without changing the inputs or the code, when I run the program multiple times, sometimes it will fail on the `date-time` validation and sometimes it would fail on the `email` one.

ssciolla replied:
1) My understanding is that JSON doesn't have those data types (https://www.json.org/json-en.html, https://json-schema.org/understanding-json-schema/reference/type.html). Do you have a different understanding?

2) I'm not sure why you're getting different ordering there. I'll do a little investigation. This doesn't seem like a big issue to me though; you fix one error and then you fix the next.

ssciolla replied:
Oh, sorry, I'm now seeing what you're referring to here: https://json-schema.org/understanding-json-schema/reference/string.html#format

I'll take a look.

ssciolla replied:
Following up to my direct message. JSON Schema supports a `format` keyword in schemas, but it seems not all implementations support it. The `jsonschema` Python package does seem to support it optionally; see the docs here: https://python-jsonschema.readthedocs.io/en/latest/validate/#validating-formats

It looks like you may need to modify the `validate` call in `environ.py` and install another package to validate `date-time`. Once everything is in place, I think you should be able to do something like this:

```
{
   "type": "string",
   "format": "date-time"
}
```

lsloan replied:
I've resolved this with commit fa3ec24.

lsloan commented:
@ssciolla, Zhen suggested I get this merged soon so I can move onto other tasks.  So, some of questions for you:

* You've made comments on this PR, but haven't given an explicit approval.  (At least, GH doesn't show it as such.  I could swear you'd approved.  🤷)  Do you have time to give an approval soon?
* If you approve this PR, is that sufficient for merging, or would you like to wait for Matt's approval, too, before I merge it?
* I mention in the description of this PR that #143 should be merged first.  However, it may be more expedient to combine that PR with this one.  What do you think?

ssciolla commented:
> @ssciolla, Zhen suggested I get this merged soon so I can move onto other tasks. So, some of questions for you:
> 
>     * You've made comments on this PR, but haven't given an explicit approval.  (At least, GH doesn't show it as such.  I could swear you'd approved.  🤷)  Do you have time to give an approval soon?
> 
>     * If you approve this PR, is that sufficient for merging, or would you like to wait for Matt's approval, too, before I merge it?
> 
>     * I mention in the description of this PR that #143 should be merged first.  However, it may be more expedient to combine that PR with this one.  What do you think?

I'm sorry, @lsloan, I've been a little preoccupied the last couple days preparing/responding to the release. Here are answers to your questions:

1) I have not approved this PR yet. I will review this thoroughly first thing on Monday.

2) I think the one approval is sufficient for merging. I tend to prefer multiple approvals, particularly when we have no automated testing, but I've been chastised before for waiting too long to merge my own PRs. One approval will be enough here in my opinion if the approve has run your code and reviewed the data for reasonableness. You can always add more reviewers if some people aren't responding.

3) I have not had time to dig into your other PR (#143), though I've had some passing thoughts about the name of the class (would prefer `DataSourceStatus` like the table name; mentioning `Timestamp` is confusing to me, since that's just a data element). I'm in favor of waiting to review/merge that PR after this one; it's not urgent or a necessary part of this Issue/PR. Also, I don't think you need to worry about backwards compatibility or changing other people's code. If you think this is a better way to do it, we change it everywhere.

ssciolla commented on db/migrations/0015.add_mivideo_creation_tables.py:
```diff
@@ -0,0 +1,28 @@
+        ) ENGINE = InnoDB DEFAULT CHARSET = utf8mb4;
+    '''),
+    step('''
+        CREATE TABLE IF NOT EXISTS mivideo_media_courses (
```
Also, did you give any thought to an auto-incremented PK instead of the compound key? Maybe there was a concern related to the `_queryRunner` part of this?

lsloan replied:
Name: I could possibly see naming it `mivideo_media_placement` or `_relationship`.  However, it **_is_** supposed to show the relationship between the `mivideo_media_creation` table and the `course` table.  Also, if we are worried about the name of the table, shouldn't we worry about the name of the `course_id` column?

Key: I wanted the table's `media_id`, `course_id`, `in_context` triple to be unique.  I was under the impression that in MySQL, making the columns part of a compound key was the right (or a good) way to do that.  If we can make MySQL enforce that uniqueness without the columns being the key, then I could go along with making the key an auto-incremented ID instead.  If you can point me to the documentation that shows how to do that, I could do that.

ssciolla replied:
Name: What do you mean about being worried about `course_id`? That's an unenforced foreign key; all other foreign keys to the `course` table use `course_id`. Just in general, I think it's useful to understand what you are modeling with a table; what does one record represent? For example, I would prefer `enrollment` over `course_user`. I'm not insisting on you changing this; you can change or resolve based on what you think would be least confusing to others.

lsloan replied:
I didn't think you were insisting on changing this.  I was trying to understand the logic of your argument.  I think for now, we'll leave this as is.  I'll open an issue to think about it again later.

I hope I know enough, but I'm always happy to learn more.  I had wanted to use an autoincrementing ID key, but I couldn't find a way to do that and keep these triples unique.

lsloan commented on README.md:
```diff
@@ -88,7 +88,10 @@ a more lenient and customizable variant of JSON. Complete the following items in
-    `MIVIDEO` | `service_account_json_filename` | The name of the JSON credential file for accessing UDP's Google BigQuery service account.  This file name is appended to the value of `ENV_DIR` (which is `/config/secrets`, by default) to determine the full path to the file.<br/><br/>If this key's value is set to `umich-its-tl-reports-prod.json` and `ENV_DIR` has its default value, the full path t ...
+    `MIVIDEO` | `udp_service_account_json_filename` | The name of the JSON credential file for accessing UDP's Google BigQuery service account.  It should be the `umich-its-tl-reports-prod.json` credential file from TL's security files folder in Box.  This file name is appended to the value of `ENV_DIR` (which is `/config/secrets`, by default) to determine the full path to the file.<br/><br/>If t ...
+    `MIVIDEO` | `default_last_timestamp` | The MiVideo procedures use the last timestamp found in its tables in this application's DB to query for data newer than that time.  If that timestamp isn't found (e.g., the first time the application runs) the value of this property will be used.  This must be a valid ISO 8601 timestamp in the UTC time zone.  The recommended value is `2020-03-01T00:00:00 ...
+    `MIVIDEO` | `kaltura_partner_id` | This is an integer that represents the Kaltura account number.  This value can be found in TL's security files folder in Box.
```
OK.  I agree with those reasons.  I'll remove the name of the folder/service the file is kept in, but I'll retain the name of the file.  I'll also mention that the filename is for the UMich-ITS-TL group.

lsloan commented on config/env_blank.hjson:
```diff
@@ -29,9 +29,13 @@
+        "default_last_timestamp": "2020-03-01T00:00:00+00:00",
+        "udp_service_account_json_filename": "umich-its-tl-reports-prod.json",
+        "kaltura_partner_id": -1,
+        "kaltura_user_secret": "kaltura_authentication_secret_here"
```
Will the JSON schema validation give the user an error if they don't fill in an empty string?  It should.

If that validation works, and because the parameters are described in `README.md`, I could remove them.

Since HJSON allows for comments, how do you feel about using them here?

ssciolla replied:
Right now, it won't give an error with an empty string, but there are some ways to specify lengths: see [here](https://json-schema.org/understanding-json-schema/reference/string.html#length). We may want to open this as a separate issue. I just started with type checking the structure and variables and adding some enumerations and required properties.

I'm trying to keep the sample HJSON relatively uncluttered, and let people use comments in their own versions. That said, we may consider moving the field descriptions out of the `README.md` and into the HJSON at some point. I'd wait on it though, personally.

lsloan replied:
Yeah, I noticed `minLength` in the documentation after I asked about it.  If it works right out of the box, I'll add that here.

lsloan commented on config/env_blank.hjson:
```diff
@@ -29,9 +29,13 @@
+        "kaltura_user_secret": "kaltura_authentication_secret_here"
     },
 
+
```
Usually my editor will remove those from file types it knows.  I think it doesn't know how to treat HJSON files.

lsloan commented on config/env_schema.hjson:
```diff
@@ -70,9 +70,16 @@
+                "default_last_timestamp": {"type": "string", "format": "date-time"},
+                "udp_service_account_json_filename": {"type": "string"},
+                "kaltura_partner_id": {"type": "integer"},
+                "kaltura_user_secret": {"type": "string"},
```
True.  But I like it there and at the end of Python data structures, too.  It may prevent an error or two from someone forgetting to add a comma when they add a new line.

ssciolla replied:
Just to be clear, I'm not asking or requiring you to change this. In my own usage (unless I make a mistake), I choose to only include commas if they are necessary and meaningful and not worry about future-proofing. I will leave what to do with this up to you.

lsloan commented on db/migrations/0015.add_mivideo_creation_tables.py:
```diff
@@ -0,0 +1,28 @@
+
+steps = [
+    step('''
+        CREATE TABLE IF NOT EXISTS mivideo_media_created (
```
The intention for the table was to store data about when media was created and with which Canvas course is it related.  There are several other attributes of the media available, but we're not capturing them here.  If we want to be really strict here, we could:

1. Keep this name and remove the `duration` column, because that's not strictly related to the creation, identity, or LMS relationship of the media.
1. Change the name to `mivideo_media` and add all the various attributes of the media that could possibly be of use to us now or in the near future.

Probably not surprisingly, I don't think we need to change anything here.

ssciolla replied:
I think the answer to this will depend on what the business question is that we are trying to answer. I think I'm in favor of `mivideo_media` if you keep `duration`; I don't know whether that element is needed or not.

lsloan commented on mivideo/mivideo_extract.py:
```diff
@@ -33,32 +41,42 @@ class MiVideoExtract(object):
+        return udpDb
 
-        self.appDb: DBCreator = DBCreator(dbParams)
+    def _kalturaInit(self):
+        self.kPartnerId = self.mivideoConfig['kaltura_partner_id']
+        self.kUserSecret = self.mivideoConfig['kaltura_user_secret']
```
They only need to have values when the Kaltura API code runs.  So if that code doesn't run, I thought it would be better to leave it out.

Normally, the UDP and the Kaltura parts of this code always run whenever this MiVideo procedure is run by the main program.  However, as I was coding and testing, I would sometimes comment out one or the other in the `run()` method.  I can imagine that in the future, we might be able to tell the main program to run specific parts of a module.  (Or maybe it would be better to separate the MiVideo code into separate UDP and Kaltura modules.)

ssciolla replied:
Hm, I see. Yeah, separating at some point might make sense, but not required at this point. I also don't see a huge problem with setting them in `__init__` and then not using them.

lsloan commented on mivideo/mivideo_extract.py:
```diff
@@ -67,62 +85,246 @@ def _readTableLastTime(self, tableName: str, tableColumnName: str) -> Union[date
+
+        sql = (
+            f'INSERT INTO {tableFullName} ({columnNames}) VALUES ({valuePlaceholders}) '
+            f'ON DUPLICATE KEY UPDATE {columnNameList[0]}={columnNameList[0]}'  # magic
```
It needed to do **_something_** and updating one column was the smallest something it could do.

ssciolla replied:
Gotcha, okay.

lsloan commented on mivideo/mivideo_extract.py:
```diff
@@ -67,62 +85,246 @@ def _readTableLastTime(self, tableName: str, tableColumnName: str) -> Union[date
+        lastTime: Union[datetime, None] = (
+            self._readTableLastTime(tableName, 'created_at'))
+
+        if (lastTime):
+            localLogger.info(
+                f'Last time found in table: "{lastTime.isoformat()}"')
+        else:
+            lastTime = datetime.fromisoformat(self.defaultLastTimestamp)
+            localLogger.info('Last time not found in table; '
+                             f'using default time: "{lastTime.isoformat()}"')
```
I thought about that.

I wanted to be able to use a default value and specific logging message when a value wasn't found in the table.  I didn't feel like the logging should be done in that function.  (Although I do use logging in other functions.)

I could probably compromise by passing in the default value, apply the logic in the function, and do the logging there as well.

ssciolla replied:
I think you're change is better. I don't have the same qualms about putting logging in functions like this.

lsloan commented on mivideo/mivideo_extract.py:
```diff
@@ -67,62 +85,246 @@ def _readTableLastTime(self, tableName: str, tableColumnName: str) -> Union[date
+
+        kFilter = KalturaMediaEntryFilter()
+        kFilter.createdAtGreaterThanOrEqual = createdAtTimestamp
+        kFilter.categoriesFullNameIn = 'Canvas_UMich'
```
Sure.

I notice you used the term "configuration" variable, not "environment".  Nudge, nudge.  Wink, wink.

ssciolla replied:
I had `environment` first and then edited it.

lsloan commented on mivideo/mivideo_extract.py:
```diff
@@ -67,62 +85,246 @@ def _readTableLastTime(self, tableName: str, tableColumnName: str) -> Union[date
+        queryPageNumber: int = kPager.pageIndex  # for logging purposes
+        totalNumberResults: int = numberResults  # for logging purposes
+
+        while True:
```
OK, so…

```python
trueFlag = True
while trueFlag:
  …
```

\<Ducks and runs\>

lsloan replied:
Seriously, though.  I'll give it some thought.

ssciolla replied:
I'm thinking more like `more_pages = True`. Then when you set it to `False`, people actually know you intend to break a certain loop, rather than them wondering, "does this `break` apply to that `for` or this `while`? It's just clearer code in my opinion.

ssciolla replied:
Yeah, I like `endOfResults` than `while True`.

lsloan commented on vocab.py:
```diff
@@ -22,6 +22,7 @@ class ValidDataSourceName(Enum):
+    KALTURA_API = auto()
     UNIZIN_DATA_PLATFORM_EVENTS = auto()
     UNIZIN_DATA_WAREHOUSE = auto()
     CANVAS_ZOOM_MEETINGS = auto()
```
😛

lsloan commented on mivideo/mivideo_extract.py:
```diff
@@ -67,62 +85,246 @@ def _readTableLastTime(self, tableName: str, tableColumnName: str) -> Union[date
+                lastId = results[-1].id
+                totalNumberResults += numberResults
+
+            if (numberResults < kPager.pageSize):
```
If the last page size (`numberResults`) is the same as the pager's page size, then the loop would continue for another round.  The API would return a page of zero results, which would be less than the pager's page size.

ssciolla replied:
Okay, if that's how it works, then it's probably fine. Just concerned about trying to access a page that doesn't exist, or some such.

lsloan commented on mivideo/mivideo_extract.py:
```diff
@@ -67,62 +85,246 @@ def _readTableLastTime(self, tableName: str, tableColumnName: str) -> Union[date
+
+                creationData: pd.DataFrame = self._makeCreationData(resultDictionaries)
+
+                creationData.to_sql(
+                    tableName, self.appDb.engine, if_exists='append', index=False)
+
+                courseData: pd.DataFrame = self._makeCourseData(resultDictionaries)
+
+                courseData.to_sql('mivideo_media_courses', self.appDb.engine, if_exists='append',
+                                  index=False, method=self._queryRunner)
```
Yeah, simply that I thought writing progressively would be better than waiting to get everything first, risking data loss and large transactions.  Since each page of results is independent, there's no reason to keep them together.  I also thought that it would make the calls to Pandas **_slightly_** simpler; always creating a new dataframe rather than creating one the first time, then appending thereafter.

ssciolla replied:
Okay, yeah, if the data is always different on each page (i.e. you don't need to drop duplicates ever across pages), this approach should work fine. Just curious as to your thinking. I kinda go back and forth about whether it would be worse to have an incomplete dataset (it fails halfway through) or have no data at all (it fails anytime before a final insertion).

lsloan replied:
I see.  In this case, it's fine to have an incomplete dataset.  The query always looks for anything newer than the last timestamp in the DB.  If something causes this program to fail part way through processing the API results, then the next time the program runs successfully, it will continue with the last timestamp found in the DB.

lsloan commented:
@ssciolla, thanks!

I've addressed some of the concerns you raised in the review.  Take a look and let me know if there's anything glaringly wrong.

If it's not bad, I'll merge tomorrow morning.

ssciolla commented on README.md:
```diff
@@ -88,7 +88,10 @@ a more lenient and customizable variant of JSON. Complete the following items in
+    `MIVIDEO` | `udp_service_account_json_filename` | The name of the JSON credential file for accessing UDP's Google BigQuery service account.  It should be the `umich-its-tl-reports-prod.json` credential file for UMich ITS TL.  This file name is appended to the value of `ENV_DIR` (which is `/config/secrets`, by default) to determine the full path to the file.<br/><br/>If this key's value is set ...
+    `MIVIDEO` | `default_last_timestamp` | The MiVideo procedures use the last timestamp found in its tables in this application's DB to query for data newer than that time.  If that timestamp isn't found (e.g., the first time the application runs) the value of this property will be used.  This must be a valid ISO 8601 timestamp in the UTC time zone.  The recommended value is `2020-03-01T00:00:00 ...
+    `MIVIDEO` | `kaltura_partner_id` | This is an integer that represents the Kaltura account number.  UMich ITS TL users can find this value in the usual security files folder.
+    `MIVIDEO` | `kaltura_user_secret` | This is a string that represents an administrator's key for the Kaltura account.  UMich ITS TL users can find this value in the usual security files folder.
```
You'll need to add an entry for the new configuration variable for `Canvas_UMich`.

ssciolla replied:
You'll need to add an entry for your new configuration variable for "Canvas_Umich".

ssciolla commented on mivideo/mivideo_extract.py:
```diff
@@ -67,62 +88,248 @@ def _readTableLastTime(self, tableName: str, tableColumnName: str) -> Union[date
+                    kPager.pageIndex = 1
+                    continue
+
+                localLogger.info(f'Other Kaltura API error: "{kException}"')
+                break
```
I don't quite understand the `KALTURA_MAX_MATCHES_ERROR` (what does that mean?), but have you not had any issues with other connection-related exceptions? Seems like we might need some re-try behavior?

ssciolla commented on mivideo/mivideo_extract.py:
```diff
@@ -67,62 +88,248 @@ def _readTableLastTime(self, tableName: str, tableColumnName: str) -> Union[date
+                    kPager.pageIndex = 1
+                    continue
+
+                localLogger.info(f'Other Kaltura API error: "{kException}"')
+                break
```
I don't quite understand the KALTURA_MAX_MATCHES_ERROR (what does that mean?), but have you not had any issues with other connection-related exceptions? Seems like we might need some re-try behavior?

lsloan replied:
Kaltura allows exactly 10,000 results from any API query.  If you attempt to read past that, they throw a `KalturaException` with a message that contains the string `QUERY_EXCEEDED_MAX_MATCHES_ALLOWED`.  Unfortunately, they don't throw a specific exception for this condition.  They use that general one with a string.

The way to handle that error is to change the filter and query again.  In our case, since we're querying for things newer than a specific time and they're sorted increasingly, we can make a new query by using the last time received in the API's results.  When we use a new time in subsequent API queries, we also must exclude the last item received by ID, because their filter is for `createdAtGreaterThanOrEqual`.  They don't have a strictly equal filter.  🙁 

So, yes, this **_is_** retry behavior.  I've not seen other connection problems, but I've also not tried to induce them.  We may need handling for that.  I'd suggest that should be a new issue.

ssciolla replied:
I think we do need handling for other exceptions, but I am okay with you creating a separate issue, so long as this is handled in the near future and not postponed indefinitely. If what we're going through with Canvas right now can be extrapolated, we cannot assume that API endpoints will always perform perfectly.

lsloan replied:
Opened issue #148.

ssciolla commented on README.md:
```diff
@@ -88,7 +88,10 @@ a more lenient and customizable variant of JSON. Complete the following items in
+    `MIVIDEO` | `udp_service_account_json_filename` | The name of the JSON credential file for accessing UDP's Google BigQuery service account.  It should be the `umich-its-tl-reports-prod.json` credential file for UMich ITS TL.  This file name is appended to the value of `ENV_DIR` (which is `/config/secrets`, by default) to determine the full path to the file.<br/><br/>If this key's value is set ...
+    `MIVIDEO` | `default_last_timestamp` | The MiVideo procedures use the last timestamp found in its tables in this application's DB to query for data newer than that time.  If that timestamp isn't found (e.g., the first time the application runs) the value of this property will be used.  This must be a valid ISO 8601 timestamp in the UTC time zone.  The recommended value is `2020-03-01T00:00:00 ...
+    `MIVIDEO` | `kaltura_partner_id` | This is an integer that represents the Kaltura account number.  UMich ITS TL users can find this value in the usual security files folder.
+    `MIVIDEO` | `kaltura_user_secret` | This is a string that represents an administrator's key for the Kaltura account.  UMich ITS TL users can find this value in the usual security files folder.
```
You'll need to add an entry for your new configuration variable for "Canvas_Umich".

lsloan replied:
I meant to do that yesterday.  ಠ_ಠ
